Sprint 14

nataliyaTruf opened the pull request:
Implemented the core functionality for the Tracker app as outlined in Sprint 14. This includes:

- Main and Tracker Creation screens according to design specs
- Schedule screen for habit tracking
- Enhanced UX with dynamic tracker updates and validation logic
- Date selection and tracker completion logic integrated

Ready for review

nataliyaTruf committed:
Basic project setup for code-based UI layout

nataliyaTruf committed:
Add basic colors and images to assets

nataliyaTruf committed:
Add fonts

nataliyaTruf committed:
Layout launch Screen

nataliyaTruf committed:
Initial layout and configuration for TrackersViewController

nataliyaTruf committed:
Initial layout for StatisticsViewController

nataliyaTruf committed:
Initialize MainTabBarController with TrackersVC and StatisticsVC child VCs

nataliyaTruf committed:
Add Tracker, TrackerCategory, and TrackerRecord models for habit tracking functionality

nataliyaTruf committed:
Add SelectTrackerViewController

nataliyaTruf committed:
WIP: Initial setup of UICollectionView in TrackersViewController

nataliyaTruf committed:
Add GeometricParams struct for layout management

nataliyaTruf committed:
WIP: Implementing basic layout for TrackerCell

nataliyaTruf committed:
WIP: Implementing basic layout for TrackersHeader

nataliyaTruf committed:
WIP: Implement tracker completion logic and day count display

Added functionality to toggle tracker completion status and dynamically update the count of completed days

nataliyaTruf committed:
Implement logic to prevent marking future-dated trackers as completed in TrackersViewController

nataliyaTruf committed:
WIP: Implement date filtering in TrackersViewController to display trackers based on selected date

nataliyaTruf committed:
Implement search filtering in TrackersViewController

Users can now refine the list of displayed trackers by selecting a specific date or by entering search criteria

nataliyaTruf committed:
WIP: Implementing basic layout for CreateTrackerViewController

- Initialized CreateTrackerViewController with a UIScrollView and a vertical UIStackView.
- Integrated CustomOptionView as a reusable component for category and schedule options.
- Added placeholder constraints for future UI components and prepared the structure for further development.

Note: CustomOptionView is designed as a separate UIView subclass for flexibility and potential reuse across the app

nataliyaTruf committed:
Implement functionality for cancel and create buttons on the tracker creation screen

Added the logic to dismiss the tracker creation screen (CreateTrackerViewController) without making any changes upon tapping the "Отменить" button and to add the newly created tracker to the main list and display it on the main screen upon tapping the "Создать" button

nataliyaTruf committed:
Implement schedule screen for tracker habit customization

- Construct ScheduleViewController to manage user interactions for setting up a habit's weekly recurrence
- Utilize a UITableView within ScheduleViewController, populated with DayTableViewCell instances for each day of the week
- Equip each DayTableViewCell with a UISwitch, enabling the user to designate whether the habit occurs on that day, indicated by the switch's 'on' state
- Embed logic in ScheduleViewController to respond to UISwitch toggles in DayTableViewCell, updating the ReccuringSchedule model to reflect user choices
- Synchronize UISwitch states with the current ReccuringSchedule to ensure visual consistency and user input retention across sessions

nataliyaTruf committed:
Refactor: Extract mock data and minor code improvements

- Extract mock data to dedicated MockDataService for better reusability and maintainability
- Implement MARK comments for improved code navigation and readability
- Adjust minor UI setup methods for consistency across the project

nataliyaTruf committed:
Add application icon for an improved visual appearance

Helvetii951 commented on Tracker/Controllers/ScheduleViewController.swift:
```diff
@@ -0,0 +1,192 @@
+    
+    // MARK: - Properties
+    
+    var onScheduleUpdated: ((ReccuringSchedule) -> Void)?
+    var schedule = ReccuringSchedule(mondays: false, tuesdays: false, wednesdays: false, thursdays: false, fridays: false, saturdays: false, sundays: false)
+    let days = ["Понедельник", "Вторник", "Среда", "Четверг", "Пятница", "Суббота", "Воскресенье"]
```
В проекте есть enum WeekDay. Было бы лучше работать с массивом значений этого енума при работе с расписанием.

Helvetii951 commented on Tracker/Controllers/CreateTrackerViewController.swift:
```diff
@@ -0,0 +1,314 @@
+// MARK: - Utility Methods
+
+extension CreateTrackerViewController {
+    private func formatScheduleText(schedule: ReccuringSchedule) -> String {
+        var days: [String] = []
+        if schedule.mondays { days.append("Пн") }
+        if schedule.tuesdays { days.append("Вт") }
+        if schedule.wednesdays { days.append("Ср") }
+        if schedule.thursdays { days.append("Чт") }
+        if schedule.fridays { days.append("Пт") }
+        if schedule.saturdays { days.append("Сб") }
+        if schedule.sundays { days.append("Вс") }
+        return days.joined(separator: ", ")
+    }
```
Раз уж было решено сделать расписание Bool'ами, рекомендую формирование этого текста занести в `ReccuringSchedule` и сделать вычисляемым свйоством.

Helvetii951 commented on Tracker/Controllers/ScheduleViewController.swift:
```diff
@@ -0,0 +1,192 @@
+    }
+    
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: DayTableViewCell.dayCellIdentifier, for: indexPath) as? DayTableViewCell else {
+            fatalError("Unable to dequeue DayTableViewCell")
+        }
```
Рекомендую избегать `fatalError`. Ее плохая сторона в том, что эта функция может уронить приложение в релизном билде.

Можно использовать [assertionFailure](https://developer.apple.com/documentation/swift/assertionfailure(_:file:line:)?changes=latest__7). А из функции возвращать `UITableViewCell()`.

Helvetii951 commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,355 @@
+        let datePicker = UIDatePicker()
+        datePicker.widthAnchor.constraint(equalToConstant: 120).isActive = true
+        datePicker.datePickerMode = .date
+        datePicker.preferredDatePickerStyle = .compact
+        datePicker.locale = Locale(identifier: "ru_RU")
+        datePicker.addTarget(
```
Не рекомендую привязывать форматтеры к конкретной локали. Пусть текст иногда будет расходиться с дизайном, но пользователь увидит привычное отображение даты.

nataliyaTruf replied:
забавно мне в прошлом проекте текст на дате из-за локали посчитали критическим замечанием, в этот раз перестраховалась. Спасибо за пояснения

Helvetii951 commented on Tracker/Controllers/TrackersViewController.swift:
```diff
@@ -0,0 +1,355 @@
+        emptyStateImageView.isHidden = hasTrackersToShow
+        
+        if !hasTrackersToShow {
+            if isSearching {
+                emptyStateLabel.text = "Ничего не найдено"
+                emptyStateImageView.image = UIImage(named: "error2")
```
Здесь можно использовать [тернарный оператор](https://www.programiz.com/swift-programming/ternary-conditional-operator)
```
emptyStateLabel.text = isSearching ? "Ничего не найдено": "Что будем отслеживать?"
```

Helvetii951 commented on Tracker/Models/Tracker.swift:
```diff
@@ -0,0 +1,16 @@
+    let name: String
+    let color: String
+    let emodji: String
+    let scedule: ReccuringSchedule?
```
Опечатка. `scedule` -> `schedule`.

nataliyaTruf committed:
Post-review enhancements for scheduling and UI clarity

- Refactored RecurringSchedule and Weekday for improved readability and localization support
- Updated controllers to streamline scheduling and enhance localization
- Partially replaced fatalError with assertionFailure for safer error handling in select scenarios
- Adjusted datePicker width and removed specific locale setting to support broader localization
- Streamlined empty state logic with ternary operators for improved UI clarity
